### ISSUE TITLE
Sphinx 7.2.0 problems with ``pathlib.Path`` instead of string paths

### ISSUE THREAD

**Issue by tronical** [sphinx-doc/sphinx#11605]
### Describe the bug

When using breathe with the 7.2.0 release, we get this error:

```
      File "/Users/simon/.local/share/virtualenvs/docs-SEK-9aTg/lib/python3.11/site-packages/breathe/project.py", line 116, in __init__
        self._default_build_dir = os.path.dirname(app.doctreedir.rstrip(os.sep))
                                                  ^^^^^^^^^^^^^^^^^^^^^
    AttributeError: 'PosixPath' object has no attribute 'rstrip'
    The full traceback has been saved in /var/folders/x6/97fjm0c540s7yn9zlv7fp9k00000gn/T/sphinx-err-sbf8ny4n.log, if you want to report the issue to the developers.
    Please also report this if it was a user error, so that a better error message can be provided next time.
    A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

I suspect that breathe relied on `doctreedir` being a string, and now it's a `PosixPath`.

If you believe that this is in error and that breathe should be changed instead to not rely on this, let me know and I'll file an issue there :)

### How to Reproduce

```bash
$ echo "Test" > index.rst
$ echo "extensions = ['breathe']" > conf.py
$ sphinx-build -M html . _build
Running Sphinx v7.2.0

Exception occurred:
  File "/Users/simon/.local/share/virtualenvs/fob-HUh8Hd7H/lib/python3.11/site-packages/breathe/project.py", line 116, in __init__
    self._default_build_dir = os.path.dirname(app.doctreedir.rstrip(os.sep))
                                              ^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'PosixPath' object has no attribute 'rstrip'
The full traceback has been saved in /var/folders/x6/97fjm0c540s7yn9zlv7fp9k00000gn/T/sphinx-err-0d2v5eaa.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

### Environment Information

```text
Platform:              darwin; (macOS-13.5-x86_64-i386-64bit)
Python version:        3.11.4 (main, Jun 20 2023, 16:59:59) [Clang 14.0.3 (clang-1403.0.22.14.1)])
Python implementation: CPython
Sphinx version:        7.2.0
Docutils version:      0.20.1
Jinja2 version:        3.1.2
Pygments version:      2.16.1
```

### Sphinx extensions

```python
['breathe']
```

**Comment by jmckenna** [comment]
@picnixz Have you filed this in the Breathe repo as an issue?  I have found them very receptive to any reports that I have made there in the past regarding Sphinx versions support.  Please report as an issue at https://github.com/breathe-doc/breathe/issues  Thanks

**Comment by stephenfin** [comment]
I think this should be reopened. This is a breaking API change that affects multiple plugins (see: https://review.opendev.org/c/openstack/openstackdocstheme/+/891678, https://github.com/click-contrib/sphinx-click/actions/runs/5893388282/job/15984823090). At a minimum, this should have pushed to a major release. Better, there should have been a translation shim and a deprecation period.

Can we revert https://github.com/sphinx-doc/sphinx/commit/49d830467098cc149390bc8349c19b7501e0920f and wait until 8.0 to re-merge this, please?

**Comment by AA-Turner** [comment]
Part of the challenge is that until recently, ``Path`` couldn't be subclassed, and so creating the deprecation pass-through would be fairly challenging. The other is that you extensions should really be using [``os.fspath``](https://docs.python.org/3/library/os.html#os.fspath), though I appreciate that as Sphinx used string paths internally the need perhaps was not felt.

A

**Comment by lucyleeow** [comment]
Just a query, it seems that `BuildEnvironment.docname` can now be `None`:

https://github.com/sphinx-doc/sphinx/blob/ed84d63e6f2c4fd43b97fc43ee8be4156a13af9e/sphinx/environment/__init__.py#L545

or at least `env.docname` in `read-source` event can now be `None`:

https://github.com/sphinx-doc/sphinx/blob/da9f8a5c33ad5eeef05dd780f2988f3ff7351ef3/sphinx/io.py#L114

Is this related to the str to `Path` change?

Ref: https://github.com/sphinx-gallery/sphinx-gallery/issues/1175

**Comment by AA-Turner** [comment]
I've released [Sphinx 7.2.3](https://pypi.org/project/Sphinx/7.2.3/) which should fix these issues.

A
